File: frontend/src/test/quantities.test.ts
import { describe, expect, it } from 'vitest'

import {
  convertFromSI,
  formatQuantity,
  quantityForSIUnit,
  unitOption,
} from '@/lib/quantities'

describe('quantity conversions', () => {
  // SignalK is SI-canonical, so every one of these is a conversion the old
  // two-function units.ts could not do.
  it('converts oil pressure out of pascals', () => {
    // 35 psi, a plausible warm idle.
    expect(convertFromSI(241_325, 'pressure', 'psi')).toBeCloseTo(35, 1)
    expect(convertFromSI(241_325, 'pressure', 'kPa')).toBeCloseTo(241.3, 1)
    expect(convertFromSI(241_325, 'pressure', 'bar')).toBeCloseTo(2.41, 2)
  })

  it('converts coolant temperature out of kelvin', () => {
    expect(convertFromSI(361.15, 'temperature', 'C')).toBeCloseTo(88, 2)
    expect(convertFromSI(361.15, 'temperature', 'F')).toBeCloseTo(190.4, 1)
  })

  it('converts fuel rate out of cubic metres per second', () => {
    // 20 L/h.
    const litresPerHourInSI = 20 / 1000 / 3600
    expect(convertFromSI(litresPerHourInSI, 'volumetricFlow', 'Lph')).toBeCloseTo(20, 3)
    expect(convertFromSI(litresPerHourInSI, 'volumetricFlow', 'gph')).toBeCloseTo(5.28, 2)
  })

  it('converts engine hours out of seconds', () => {
    expect(convertFromSI(3600 * 1234, 'duration', 'h')).toBeCloseTo(1234, 6)
  })

  it('converts revolutions out of hertz, because tachometers show RPM', () => {
    expect(convertFromSI(30, 'frequency', 'rpm')).toBeCloseTo(1800, 6)
  })

  it('converts a tank level ratio to a percentage', () => {
    expect(convertFromSI(0.8106, 'ratio', 'percent')).toBeCloseTo(81.06, 4)
  })

  it('converts tank capacity out of cubic metres', () => {
    expect(convertFromSI(1.2, 'volume', 'L')).toBeCloseTo(1200, 6)
  })
})

describe('quantity inference', () => {
  it('infers the quantity from SignalK meta units', () => {
    expect(quantityForSIUnit('Pa').id).toBe('pressure')
    expect(quantityForSIUnit('K').id).toBe('temperature')
    expect(quantityForSIUnit('m3/s').id).toBe('volumetricFlow')
    expect(quantityForSIUnit('Hz').id).toBe('frequency')
  })

  it('falls back to unitless for an unknown or missing unit', () => {
    expect(quantityForSIUnit('parsecs').id).toBe('raw')
    expect(quantityForSIUnit(undefined).id).toBe('raw')
  })

  it('falls back to the first unit when the configured one is unknown', () => {
    // A config written against a unit that later disappears must still render.
    expect(unitOption('pressure', 'furlongs').id).toBe('kPa')
  })
})

describe('formatting', () => {
  it('uses the unit default decimals unless overridden', () => {
    expect(formatQuantity(241_325, 'pressure', 'psi')).toBe('35.0')
    expect(formatQuantity(241_325, 'pressure', 'psi', 2)).toBe('35.00')
  })

  // A gauge showing 0 when it means "no data" is the dangerous failure, so
  // absent values return null and callers render the structural dash.
  it('returns null rather than zero for absent data', () => {
    expect(formatQuantity(null, 'pressure', 'psi')).toBeNull()
    expect(formatQuantity(undefined, 'pressure', 'psi')).toBeNull()
    expect(formatQuantity(Number.NaN, 'pressure', 'psi')).toBeNull()
  })

  it('formats a real zero as zero, not as absent', () => {
    expect(formatQuantity(0, 'volumetricFlow', 'Lph')).toBe('0.0')
  })
})

// SignalK publishes electrical readings in W, V and A. Without these a
// generator's output rendered as a bare number with no unit at all.
describe('electrical quantities', () => {
  it('formats power in kilowatts', () => {
    expect(formatQuantity(7564, 'power', 'kW')).toBe('7.56')
    expect(formatQuantity(13500, 'power', 'kW', 1)).toBe('13.5')
    expect(formatQuantity(7564, 'power', 'W')).toBe('7564')
  })

  it('passes volts and amps through', () => {
    expect(formatQuantity(232, 'potential', 'V', 0)).toBe('232')
    expect(formatQuantity(39, 'current', 'A', 0)).toBe('39')
  })

  it('infers each from SignalK meta units', () => {
    expect(quantityForSIUnit('W').id).toBe('power')
    expect(quantityForSIUnit('V').id).toBe('potential')
    expect(quantityForSIUnit('A').id).toBe('current')
  })

  // Absent stays absent, the same rule every other quantity follows.
  it('formats an absent electrical reading to null', () => {
    expect(formatQuantity(null, 'power', 'kW')).toBeNull()
  })
})

describe('fuel economy', () => {
  // 792950.7 m/m³ is what the vessel published at 10.21 kn burning 23.9 L/h on
  // that engine, which works out to 0.427 nm/L the long way round.
  it('converts metres per cubic metre to nautical miles per litre', () => {
    expect(formatQuantity(792950.7, 'fuelEconomy', 'nmpl')).toBe('0.43')
  })

  it('converts to nautical miles per US gallon', () => {
    expect(formatQuantity(792950.7, 'fuelEconomy', 'nmpg')).toBe('1.62')
  })

  it('infers the quantity from SignalK meta units', () => {
    expect(quantityForSIUnit('m/m3').id).toBe('fuelEconomy')
  })
})

File: backend/quantities_test.go
package main

import (
	"math"
	"testing"
)

// The values here are deliberately the same ones frontend/src/test/quantities.test.ts
// asserts on. The two unit tables cannot drift without one of these failing.
func TestConvertToSI(t *testing.T) {
	cases := []struct {
		name     string
		quantity string
		unit     string
		display  float64
		si       float64
	}{
		{"psi to pascals", "pressure", "psi", 35.0, 241316.495},
		{"kPa to pascals", "pressure", "kPa", 241.325, 241325},
		{"bar to pascals", "pressure", "bar", 2.41325, 241325},
		{"pascals are already SI", "pressure", "Pa", 241325, 241325},
		{"celsius to kelvin", "temperature", "C", 25.0, 298.15},
		{"fahrenheit to kelvin", "temperature", "F", 77.0, 298.15},
		{"kelvin is already SI", "temperature", "K", 298.15, 298.15},
		{"litres per hour to cubic metres per second", "volumetricFlow", "Lph", 20.0, 20.0 / 1000 / 3600},
		{"hours to seconds", "duration", "h", 1204, 1204 * 3600},
		{"minutes to seconds", "duration", "min", 90, 5400},
		{"percent to ratio", "ratio", "percent", 75, 0.75},
		{"litres to cubic metres", "volume", "L", 1000, 1.0},
		{"knots to metres per second", "speed", "kn", 1.943844, 1.0},
		{"feet to metres", "length", "ft", 3.28084, 1.0},
		{"rpm to hertz", "frequency", "rpm", 1800, 30},
		{"hertz is already SI", "frequency", "Hz", 30, 30},
		{"unitless passes through", "raw", "raw", 42, 42},
	}

	for _, tc := range cases {
		got, err := convertToSI(tc.display, tc.quantity, tc.unit)
		if err != nil {
			t.Fatalf("%s: unexpected error: %v", tc.name, err)
		}
		if math.Abs(got-tc.si) > math.Abs(tc.si)*1e-6+1e-9 {
			t.Fatalf("%s: expected %v, got %v", tc.name, tc.si, got)
		}
	}
}

// An unknown quantity or unit must be an error rather than a silent
// pass-through: a zone threshold left in psi but compared against pascals
// produces an alarm that never fires, and nothing says why.
func TestConvertToSIRejectsUnknownUnits(t *testing.T) {
	if _, err := convertToSI(1, "pressure", "furlongs"); err == nil {
		t.Fatal("expected an unknown unit to be rejected")
	}
	if _, err := convertToSI(1, "spiciness", "scoville"); err == nil {
		t.Fatal("expected an unknown quantity to be rejected")
	}
}

// Round-tripping catches an inverse that was written backwards.
func TestConvertToSIRoundTripsFromSI(t *testing.T) {
	for _, q := range siQuantities {
		for _, u := range q.Units {
			si, err := convertToSI(u.FromSI(1234.5), q.ID, u.ID)
			if err != nil {
				t.Fatalf("%s/%s: unexpected error: %v", q.ID, u.ID, err)
			}
			if math.Abs(si-1234.5) > 1e-6 {
				t.Fatalf("%s/%s: round trip gave %v, want 1234.5", q.ID, u.ID, si)
			}
		}
	}
}

// SignalK publishes electrical readings in volts, amps and watts. Without
// these a generator's output renders as a bare number with no unit.
func TestConvertToSIElectricalQuantities(t *testing.T) {
	cases := []struct {
		name     string
		quantity string
		unit     string
		display  float64
		si       float64
	}{
		{"kilowatts to watts", "power", "kW", 13.5, 13500},
		{"watts are already SI", "power", "W", 7564, 7564},
		{"volts are already SI", "potential", "V", 232, 232},
		{"amps are already SI", "current", "A", 39, 39},
	}

	for _, tc := range cases {
		got, err := convertToSI(tc.display, tc.quantity, tc.unit)
		if err != nil {
			t.Fatalf("%s: unexpected error: %v", tc.name, err)
		}
		if math.Abs(got-tc.si) > math.Abs(tc.si)*1e-9+1e-9 {
			t.Fatalf("%s: expected %v, got %v", tc.name, tc.si, got)
		}
	}
}

/*
SignalK publishes propulsion.<id>.fuel.economy in metres per cubic metre.

Verified against a live vessel rather than read off a spec: at 10.21 kn with
that engine burning 23.9 L/h, SOG over burn gives 0.4272 nm/L and the published
792950.7 converts to 0.4281 — a 0.2% match, which settles both the unit and
that the figure is per-engine, not per-vessel.
*/
func TestConvertToSIFuelEconomy(t *testing.T) {
	cases := []struct {
		name    string
		unit    string
		display float64
		si      float64
	}{
		{"nautical miles per litre", "nmpl", 0.4281, 792841.2},
		{"nautical miles per US gallon", "nmpg", 1.6205, 792824.0},
		{"metres per cubic metre are already SI", "m/m3", 792950.7, 792950.7},
	}

	for _, tc := range cases {
		got, err := convertToSI(tc.display, "fuelEconomy", tc.unit)
		if err != nil {
			t.Fatalf("%s: unexpected error: %v", tc.name, err)
		}
		if math.Abs(got-tc.si) > 200 {
			t.Fatalf("%s: expected about %v, got %v", tc.name, tc.si, got)
		}
	}
}
